cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -92,7 +92,23 @@ type Etcd struct {
 	DataDir   string
 	ExtraArgs map[string]string
 	// Image specifies which container image to use for running etcd. If empty, automatically populated by kubeadm using the image repository and default etcd version
-	Image string
+	Image      string
+	SelfHosted *SelfHostedEtcd
+}
+
+// SelfHostedEtcd describes options required to configure self-hosted etcd
+type SelfHostedEtcd struct {
+	// CertificatesDir represents the directory where all etcd TLS assets are stored. By default this is
+	// a dir names "etcd" in the mains CertificatesDir value.
+	CertificatesDir string
+	// The service IP of the etcd cluster
+	ClusterIP string
+	// The number of members that will exist in the etcd cluster.
+	ClusterSize int
+	// The version of etcd running in the cluster.
+	EtcdVersion string
```

Comment:
Declaring explicitly etcd version is a kind of change of behaviour, because up to now, if I remember well,  etcd version is linked to the version of the controlplane. Now, I not totally sure that it safe to have those two versions tight together for the whole lifecycle of the cluster, but I'm somehow Warfield of creating possible version skew issues. Any thoughts here?

Reply:
Yeah, good point. TBH I'm not sure what the supported versions of etcd are, I see it's defined [here](https://github.com/kubernetes/kubernetes/blob/028ee090f64df569db3e64c8bdd9e74d08480ed2/hack/lib/etcd.sh#L19). It'd be great if k8s has some kind of compatibility matrix. 

Perhaps we could verify that the provided value is above `3.1.10`? If an invalid value is provided, the operator should catch that IMO.

Reply:
in kubeadm currently is a constant [here](https://github.com/kubernetes/kubernetes/blob/d118e44320a5084407554f2b0f08bf50f0569aba/cmd/kubeadm/app/constants/constants.go#L147)